### ISSUE TITLE
VS Code does not properly handle dfonts (bitmap fonts) on Mac

### ISSUE THREAD

**Issue by LittleGreenFrog** [microsoft/vscode#109938]
Issue Type: <b>Bug</b>

I tried with extensions turned off. It made no difference.

1. Install a Mac dfont such as gohufont 14 (https://font.gohu.org)
2. Open settings
3. Set Editor Font Family to gohufont
4. Set Editor Font Size to 14
5. Close VS Code
6. Open VS Code
7. Open a text file
8. Note that the characters are spread way out.
Expected: They should not be spread way out.
9. Open settings
10. Enter 'spacing' 
11. Set 'spacing' to -0.15
12. Exit and re-enter VS Code
13. Spacing *may* be correct. Sometimes it is still spread way out.
14. Exit and re-enter VS Code again.
15. Note: The characters are spread way out again.
Expected: VS Code should remember its settings.

VS Code version: Code 1.50.1 (d2e414d9e4239a252d1ab117bd7067f125afd80a, 2020-10-13T14:53:05.704Z)
OS version: Darwin x64 19.6.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz (16 x 2300)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|32.00GB (4.69GB free)|
|Process Argv|--crash-reporter-id 0989bf7c-d742-4a86-badc-7de99d73c54b|
|Screen Reader|no|
|VM|0%|
</details><details><summary>Extensions (8)</summary>

Extension|Author (truncated)|Version
---|---|---
numbered-bookmarks|ale|7.2.0
vscode-quick-select|dba|0.2.9
react-native-react-redux|EQu|2.0.6
vscode-rufo|jnb|0.0.2
vscode-react-native|msj|1.1.0
deepdark-material|Nim|3.2.1
ruby|reb|0.27.0
vscode-ruby|win|0.27.0

(4 theme extensions excluded)

</details>

**Comment by alexdima** [comment]
@LittleGreenFrog VS Code is built with Electron, which is built with Chromium. The font rendering in the editor is delegated to Chromium.

I have created a test page at https://jsfiddle.net/7aqvm506/ . Can you please try opening it in Chrome vs in Firefox ?

**Comment by LittleGreenFrog** [comment]
![image](https://user-images.githubusercontent.com/26017870/98578263-bc966d80-2282-11eb-9d3a-82f505b5a267.png)

Here's a screen capture.

**Comment by LittleGreenFrog** [comment]
![image](https://user-images.githubusercontent.com/26017870/98578607-2dd62080-2283-11eb-8504-2895a9195295.png)

For comparison - gohufont @ 14

**Comment by alexdima** [comment]
@LittleGreenFrog Could you please also try to reproduce at https://vscode-web-test-playground.azurewebsites.net/ ? That is basically the vscode UI running in a browser. I believe that perhaps there was something fixed in Chrome in the meantime. Chrome is now at version 87, while vscode is still using Chromium at version 83, so perhaps there was a fix that went in Chromium in that timeframe.

**Comment by David-A-RogersHS2** [comment]
I updated chrome to Version 87.0.4280.67 (Official Build) (x86_64)

The problem was exhibited in the playground.

![image](https://user-images.githubusercontent.com/20795729/100027685-e8514180-2db2-11eb-92fb-ac09cfbdb22b.png)

**Comment by alexdima** [comment]
@David-A-RogersHS2 Can you please also try https://jsfiddle.net/c0x62bma/ (note this is a new URL). This defines the HTML to more closely resemble what vscode does:
```
<div style="font-family:gohufont; font-weight: normal; font-size: 12px; font-feature-settings: 'liga' 0, 'calt' 0; line-height: 18px; letter-spacing: 0px;">
Just some text
</div>
```

Could you please try to remove the style declarations to see which one might be causing this rendering? My guess would be the `font-feature-settings: 'liga' 0, 'calt' 0;` or `letter-spacing: 0px;`.

**Comment by LittleGreenFrog** [comment]
That site did not exhibit the problem. gohufont is designed for 14px size.
So once I changed to that and re-ran, the output looked as expected as-is.

--dr


[image: image.png]

**Comment by alexdima** [comment]
Ok, then my understanding is that to use the gohufont one must also configure `"editor.fontSize": 14` and all is well.

**Comment by David-A-RogersHS2** [comment]
This should not be closed!!!!!

This is the current version of vscode
Version: 1.51.1
Commit: e5a624b788d92b8d34d1392e4c4d9789406efe8f
Date: 2020-11-11T01:11:34.018Z
Electron: 9.3.3
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Darwin x64 19.6.0

This is what it looks like using gohufont at 14 point

Version: 1.51.1
Commit: e5a624b788d92b8d34d1392e4c4d9789406efe8f
Date: 2020-11-11T01:11:34.018Z
Electron: 9.3.3
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Darwin x64 19.6.0

This is what it looks like using gohufont at 14 point.

![image](https://user-images.githubusercontent.com/20795729/101095342-67006880-3583-11eb-91f6-28239df5a338.png)

**Comment by David-A-RogersHS2** [comment]
@alexdima Reopen this bug please.

**Comment by alexdima** [comment]
@David-A-RogersHS2 Sure. Can you please try out what I wrote in https://github.com/microsoft/vscode/issues/109938#issuecomment-737340932

**Comment by David-A-RogersHS2** [comment]
Alex - There must be some difference in the sandbox code in chrome and the font handling code in vs code. The sandbox seems to be ok even when I remove various parameters.

I did this experiment. I removed my settings.json and ran "code --disable-extensions" from the terminal. Then set the font to gohufont and the the font size to 14. As you can see, it's not working properly. This is true not only for this font but for all bitmap fonts. TT fonts work ok.

Note that this is a recent change. Older versions of VS Code have worked fine with bitmap fonts.

![image](https://user-images.githubusercontent.com/20795729/101180085-c062a900-3610-11eb-9f5c-349fce5bccf5.png)

**Comment by alexdima** [comment]
I have installed the `.dfont` variant and the first `.ttf` variant from https://font.gohu.org/ and I can confirm that for both the rendering looks like the one you show.

But then I also tried the third `.ttf` variant, from https://github.com/koemaeda/gohufont-ttf and that renders fine for me:
<img width="777" alt="Screen Shot 2020-12-04 at 17 54 38" src="https://user-images.githubusercontent.com/5047891/101191331-d6488c00-3659-11eb-96b1-92a64fbfe23b.png">

![image](https://user-images.githubusercontent.com/5047891/101191510-0b54de80-365a-11eb-9eac-7761cfee879b.png)

I am sorry I don't have an explanation for this behavior, since I am not a font expert, I can only suggest that you also try that variant of the font.

**Comment by LittleGreenFrog** [comment]
I know that ttf fonts work. But bitmapped dfonts don't. That's a
deal-breaker for me. My eyes are not as good as they used to be and bitmap
fonts are much clearer. Please pass this ticket on to someone who can fix
the problem. dfonts USED to work in VS Code.

**Comment by David-A-RogersHS2** [comment]
Do I need to mention that dfonts work in Xcode and appCode. Even emacs doesn't have this problem.

**Comment by alexdima** [comment]
@deepak1556 Do you have any ideas?

**Comment by alexdima** [comment]
>dfonts USED to work in VS Code

@LittleGreenFrog What was the last VS Code version that worked for you?

**Comment by LittleGreenFrog** [comment]
Alexandru,

I was off on other things. But I have just downloaded 1.53.2 and I am happy
to report that dfonts seem to be working again in VS Code. Thanks for
running this down.

--dr

**Comment by LittleGreenFrog** [comment]
I tried the latest version 1.53.2. And it worked the first time I loaded it up. But then when I tried again later, I was back to wide spacing.

In answer to your question. 1.48.2 will consistently display dfonts properly. 1.49.3 does not.

1.48.2 =======
![dfont_in_1 48 2](https://user-images.githubusercontent.com/26017870/108003645-0648cf80-6fb9-11eb-94fd-5e4e0fc4caa4.jpg)

1.49.3 ======

![dfont_in_1 49 3](https://user-images.githubusercontent.com/26017870/108003699-24163480-6fb9-11eb-8d96-fd26d1e72c7b.jpg)

**Comment by David-A-RogersHS2** [comment]
@alexdima @deepak1556 

See the above answer to a previous question. But also:

I don't know if this helps figure things out. But I installed the current latest stable build (1.53.2) on a Mac Big Sur machine that has never had VS Code on it before. The very first launch displayed dfonts correctly. Subsequent launches display dfonts with the super wide spacing as above. So something that happens between first and second launch is affecting dfont display.

**Comment by georgemorgan** [comment]
I am also experiencing this issue, though it is not consistent. Most of the time it renders incorrectly using either the .dfont or first .ttf GohuFont. Sometimes though it renders fine. This issue was also introduced something ~1.5 years ago. I used to use VSCode with GohuFont without issues. Using the 3rd GohuFont from their website is not an option because the scaling is incorrect.

**Comment by georgemorgan** [comment]
Atom now has this issue is well, likely caused by their most recent Electron up-rev.

**Comment by clavin** [comment]
This was reported to Electron in https://github.com/electron/electron/issues/30297 and happens to be a Chromium issue, so I reported it with more detail on what triggers it in Chromium's bug tracker at https://bugs.chromium.org/p/chromium/issues/detail?id=1235963

**Comment by deepak1556** [comment]
Closing as upstream issue has resolved the bug as wontfix. There is nothing much we can do, given the bug does not repro in upstream with latest versions https://bugs.chromium.org/p/chromium/issues/detail?id=1235963#c6, it might be fixed in latest stable `1.85`.

**Comment by David-A-RogersHS2** [comment]
Indeed. The problem is resolved in 1.85.
